core/unit_test/TestMathematicalFunctions.hpp

```diff
@@ -261,29 +285,34 @@ struct FloatingPointComparison {
 template <class>
 struct math_function_name;
 
-#define DEFINE_UNARY_FUNCTION_EVAL(FUNC, ULP_FACTOR)                  \
-  struct MathUnaryFunction_##FUNC {                                   \
-    template <typename T>                                             \
-    static KOKKOS_FUNCTION auto eval(T x) {                           \
-      static_assert(                                                  \
-          std::is_same<decltype(Kokkos::FUNC((T)0)),                  \
-                       math_unary_function_return_type_t<T>>::value); \
-      return Kokkos::FUNC(x);                                         \
-    }                                                                 \
-    template <typename T>                                             \
-    static auto eval_std(T x) {                                       \
-      static_assert(                                                  \
-          std::is_same<decltype(std::FUNC((T)0)),                     \
-                       math_unary_function_return_type_t<T>>::value); \
-      return std::FUNC(x);                                            \
-    }                                                                 \
-    static KOKKOS_FUNCTION double ulp_factor() { return ULP_FACTOR; } \
-  };                                                                  \
-  using kk_##FUNC = MathUnaryFunction_##FUNC;                         \
-  template <>                                                         \
-  struct math_function_name<MathUnaryFunction_##FUNC> {               \
-    static constexpr char name[] = #FUNC;                             \
-  };                                                                  \
+#define DEFINE_UNARY_FUNCTION_EVAL(FUNC, ULP_FACTOR)                    \
+  struct MathUnaryFunction_##FUNC {                                     \
+    template <typename T>                                               \
+    static KOKKOS_FUNCTION auto eval(T x) {                             \
+      static_assert(                                                    \
+          std::is_same<decltype(Kokkos::FUNC((T)0)),                    \
+                       math_unary_function_return_type_t<T>>::value);   \
+      return Kokkos::FUNC(x);                                           \
+    }                                                                   \
+    template <typename T>                                               \
+    static auto eval_std(T x) {                                         \
+      if constexpr (std::is_same<T, KE::half_t>::value ||               \
+                    std::is_same<T, KE::bhalf_t>::value) {              \
+        return std::FUNC(static_cast<float>(x));                        \
```

Comment:
Is it intentional to stay in floating point.
Are we comparing results as float or half types?

Reply:
Yes, this is intentional. std doesn't define any functions for for T  = half_t or T = bhalf_t. We are comparing results as float. Even when the math functions do return half_t or bhalf_t results in the future, results will be compared as float.

Reply:
I don't quite understand why comparing them as float is the right thing to do.

Reply:
@dalg24: I don't know how else to explain this. std math functions are not defined for half types. What else would you compare them to?